agent/src/datafile.rs

```diff
@@ -714,106 +756,93 @@ impl DataFile {
-
-        let paths = std::fs::read_dir(&path);
-        if paths.is_err() {
-            bail!(paths.err().unwrap());
-        }
-
-        let mut results: Vec<Snapshot> = vec![];
-        for path in paths.unwrap() {
-            if path.is_err() {
-                bail!(path.err().unwrap());
-            }
 
-            let path = path.unwrap().path();
-
-            if path.is_dir() {
-                let dir_name = path
-                    .file_name()
-                    .ok_or_else(|| {
-                        anyhow!("could not turn {:?} into filename!", path)
-                    })?
-                    .to_str()
-                    .ok_or_else(|| {
-                        anyhow!("could not turn {:?} into str!", path)
-                    })?
-                    .to_string();
-
-                let dataset = {
-                    let mut path = self.base_path.clone().to_path_buf();
-                    path.push("regions");
-                    path.push(region_id.0.clone());
-
-                    ZFSDataset::new(
-                        path.into_os_string().into_string().unwrap(),
-                    )?
-                };
-
-                let snapshot_name =
-                    format!("{}@{}", dataset.dataset(), dir_name);
-
-                // Creation time of a .zfs/snapshot/<folder> as retrieved by
-                // stat doesn't make sense. Use `zfs get`:
-                //
-                //   # zfs get -pH -o value creation <snapshot_name>
-                //   1644441276
-
-                let mut cmd = Command::new("zfs");
-                cmd.arg("get");
-                cmd.arg("-pH");
-                cmd.arg("-o");
-                cmd.arg("value");
-                cmd.arg("creation");
-                cmd.arg(snapshot_name);
+        info!(self.log, "path is {:?}", &path);
 
-                info!(
-                    self.log,
-                    "command {:?} {:?}",
-                    cmd.get_program(),
-                    cmd.get_args()
-                );
-
-                let cmd = cmd.output()?;
+        let dataset =
+            ZFSDataset::new(path.into_os_string().into_string().unwrap())?;
 
-                if !cmd.status.success() {
-                    bail!("stat didn't work!");
-                }
+        info!(self.log, "dataset is {}", dataset.dataset());
 
-                let cmd_stdout = {
-                    let cmd_stdout = String::from_utf8_lossy(&cmd.stdout);
+        let cmd = Command::new("zfs")
+            .arg("list")
+            .arg("-t")
+            .arg("snapshot")
+            .arg("-H")
+            .arg("-o")
+            .arg("name")
+            .arg("-r")
+            .arg(dataset.dataset())
+            .output()?;
 
-                    // Remove newline
-                    let cmd_stdout = cmd_stdout.trim_end().to_string();
+        if !cmd.status.success() {
+            bail!("zfs list snapshots failed!");
```

Comment:
Do we want to print out the error here if we saw one?

Reply:
yes definitely, done in dfb1358